lib/services/crawler/crawler.go

```diff
@@ -0,0 +1,153 @@
+package crawler
+
+import (
+	"database/sql"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"gorm.io/gorm"
+	"net/url"
+	"strings"
+
+	"github.com/markgravity/golang-ic/helpers/log"
+	"github.com/markgravity/golang-ic/lib/models"
+
+	"github.com/gocolly/colly/v2"
+	"github.com/gocolly/colly/v2/extensions"
+)
+
+type Crawler struct {
+	DB        *gorm.DB
+	Keyword   *models.Keyword
+	Collector *colly.Collector
+
+	parsingResult *ParsingResult
+}
+
+type ParsingResult struct {
+	HTMLCode        string
+	NonAdwordLinks  []string
+	AdwordLinks     []string
+	ShopAdwordLinks []string
+}
+
+var selectors = map[string][]string{
+	"nonAds":      {"#search .yuRUbf > a", ".ZINbbc.xpd .kCrYT > a"},
+	"ads":         {"#tads .d5oMvf > a", ".uEierd .ZINbbc a.C8nzq.BmP5tf"},
+	"mobileLinks": {".ezO2md a.fuLhoc.ZWRArf"},
+	"mobileAds":   {"span.dloBPe"},
+}
+
+const urlPattern = "https://www.google.com/search?q=%s"
+
+func (c *Crawler) Run() error {
+	if c.Keyword == nil {
+		return errors.New("keyword is required")
+	}
+
+	c.Keyword.Status = models.Processing
+	err := c.Keyword.Save(c.DB)
+	if err != nil {
+		c.Keyword.Status = models.Failed
+		_ = c.Keyword.Save(c.DB)
+		return err
+	}
+
+	parsingResult := ParsingResult{}
+	if c.Collector == nil {
+		c.Collector = colly.NewCollector()
+	}
+
+	extensions.RandomUserAgent(c.Collector)
+
+	c.Collector.OnRequest(func(r *colly.Request) {
+		log.Infof("Visiting: %v user-agent: %v", r.URL, r.Headers.Get("User-Agent"))
+	})
+
+	c.Collector.OnResponse(func(r *colly.Response) {
+		parsingResult.HTMLCode = string(r.Body[:])
+	})
+
+	c.Collector.OnError(func(r *colly.Response, err error) {
+		log.Errorf("Request URL: %v failed with error: %v", r.Request.URL, err)
+		c.Keyword.Status = models.Failed
+		_ = c.Keyword.Save(c.DB)
+	})
+
+	for _, pattern := range selectors["nonAds"] {
+		c.Collector.OnHTML(pattern, func(e *colly.HTMLElement) {
+			link := formatLink(e.Attr("href"))
+			parsingResult.NonAdwordLinks = append(parsingResult.NonAdwordLinks, link)
+		})
+	}
+
+	for _, pattern := range selectors["ads"] {
+		c.Collector.OnHTML(pattern, func(e *colly.HTMLElement) {
+			link := formatLink(e.Attr("href"))
+			parsingResult.AdwordLinks = append(parsingResult.AdwordLinks, link)
+		})
+	}
+
+	c.Collector.OnHTML(selectors["mobileLinks"][0], func(e *colly.HTMLElement) {
+		link := formatLink(e.Attr("href"))
+		if len(e.DOM.Find(selectors["mobileAds"][0]).Nodes) > 0 {
+			parsingResult.AdwordLinks = append(parsingResult.AdwordLinks, link)
+		} else {
+			parsingResult.NonAdwordLinks = append(parsingResult.NonAdwordLinks, link)
+		}
+	})
+
+	c.Collector.OnScraped(func(r *colly.Response) {
+		c.parsingResult = &parsingResult
+		err := c.save()
+		if err != nil {
+			c.Keyword.Status = models.Failed
+			_ = c.Keyword.Save(c.DB)
+			log.Error("Error when saving to DB:", err)
```

Comment:
```suggestion
			log.Error("Error when saving to DB:", err)
			c.Keyword.Status = models.Failed
			errKeyword = c.Keyword.Save(c.DB)
			
			if errKeyword != nil {
  			log.Error("Error, cannot update status to Failed:", err)

        return errKeyword
			}

      return err
``` 

I'll recommend logging the action asap after the error is identified. Maybe the update of the status will have another issue so we don't want to mix the logs :)


Ok, and should we return? I don't think we should call the parser when we could not save the keyword into the DB 🤔

Reply:
It doesn't support returning an error in `c.Collector.OnScraped(func(r *colly.Response) {` 🙏